src/libopensc/pkcs15-cac.c

```diff
@@ -334,23 +334,30 @@ static int sc_pkcs15emu_cac_init(sc_pkcs15_card_t *p15card)
 		if (cert_out->key->algorithm != SC_ALGORITHM_RSA) {
 			sc_log(card->ctx, "unsupported key.algorithm %d", cert_out->key->algorithm);
 			sc_pkcs15_free_certificate(cert_out);
+			free(pubkey_info.direct.spki.value);
 			continue;
 		} else {
 			pubkey_info.modulus_length = cert_out->key->u.rsa.modulus.len * 8;
 			prkey_info.modulus_length = cert_out->key->u.rsa.modulus.len * 8;
 			r = sc_pkcs15emu_add_rsa_pubkey(p15card, &pubkey_obj, &pubkey_info);
 			sc_log(card->ctx,  "adding rsa public key r=%d usage=%x",r, pubkey_info.usage);
-			if (r < 0)
+			if (r < 0) {
+				free(pubkey_info.direct.spki.value);
 				goto fail;
+			}
 			r = sc_pkcs15emu_add_rsa_prkey(p15card, &prkey_obj, &prkey_info);
+			if (r < 0)
+				goto fail;
```

Comment:
I think in that case, if `sc_pkcs15emu_add_rsa_pubkey` doesn't end with an error, the pubkey object with `pubkey_info.direct.spki.value` is already stored in `p15card` and then it is freed in `sc_pkcs15_unbind`. But I can fix it by directly removing the added objects with `sc_pkcs15_remove_objects` when error occurs.

Reply:
I agree that if `sc_pkcs15emu_add_rsa_pubkey` succeeds, everything is fine. But what if `sc_pkcs15emu_add_rsa_prkey` is not successful? I'm not sure about the internals of sc_pkcs15emu_add_rsa_prkey, but ideally, this function would not modify p15card until it is clear that the call succeeds. If that was the case, it would mean that prkey_info is *not* stored in p15card if sc_pkcs15emu_add_rsa_prkey returns `r < 0` and a manual call to `free(pubkey_info.direct.spki.value);` needs to be done on line 350 as well...

Reply:
Ok, I think I got your point. If `sc_pkcs15emu_add_rsa_pubkey` succeeds, the pubkey object is in `p15card`, then the ownership of `pubkey_info.direct.spki.value` is moved to that object (it should be better visible now). When `sc_pkcs15emu_add_rsa_prkey` fails, all objects on `p15card` are cleaned with `sc_pkcs15_card_clear` as well as pubkey object with the mentioned allocated buffer.
I added cleaning of the `p15card` when init fails to the rest of `sc_pkcs15emu_*_init` functions.